pkg/reconciler/namespace/namespace.go

```diff
@@ -0,0 +1,164 @@
+	logger := logging.FromContext(ctx)
+	ctx = c.configStore.ToContext(ctx)
+
+	original, err := c.nsLister.Get(key)
+
+	if apierrs.IsNotFound(err) {
+		logger.Errorf("Namespace %s in work queue no longer exists %s", key, err)
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// Don't modify the informers copy
+	newNamespace := original.DeepCopy()
+
+	err = c.reconcile(ctx, newNamespace)
+	if err != nil {
+		c.Recorder.Event(newNamespace, corev1.EventTypeWarning, "InternalError", err.Error())
+	}
+	return err
+}
+
+func (c *Reconciler) reconcile(ctx context.Context, ns *corev1.Namespace) error {
+	cfg := config.FromContext(ctx)
+
+	// Only create wildcard certs for the default domain
+	defaultDomain := cfg.Domain.LookupDomainForLabels(nil)
+	if defaultDomain == "" {
+		c.Logger.Infof("Skipping wildcard certificate creation for namespace %s. No default domain defined.", ns.Name)
+		return nil
+	}
+
+	domain, err := wildcardDomain(cfg.Network.DomainTemplate, defaultDomain, ns.Name)
+	if err != nil {
+		c.Logger.Errorf("Failed to apply domain template %s to domain %s and namespace %s: %v", cfg.Network.DomainTemplate, defaultDomain, ns.Name, err)
+		return err
+	}
+
+	desiredCert := makeDesiredCertificate(ns, domain)
+	existingCert, err := c.knCertificateLister.Certificates(ns.Name).Get(ns.Name)
+	if apierrs.IsNotFound(err) {
+		cert, err := c.ServingClientSet.NetworkingV1alpha1().Certificates(ns.Name).Create(desiredCert)
+		if err != nil {
+			c.Logger.Errorw("Failed to create namespace certificate", zap.Error(err))
+			c.Recorder.Eventf(ns, corev1.EventTypeWarning, "CreationFailed",
+				"Failed to create Knative certificate %s/%s: %v", ns.Name, desiredCert.ObjectMeta.Name, err)
+			return err
+		}
+		c.Recorder.Eventf(cert, corev1.EventTypeNormal, "Created",
+			"Created Knative Certificate %s/%s", ns.Name, cert.ObjectMeta.Name)
+	} else if err != nil {
+		return err
+	} else if !equality.Semantic.DeepEqual(existingCert.Spec, desiredCert.Spec) {
+		copy := existingCert.DeepCopy()
+		copy.Spec = desiredCert.Spec
+		_, err := c.ServingClientSet.NetworkingV1alpha1().Certificates(copy.Namespace).Update(copy)
+		if err != nil {
+			c.Logger.Errorw("Failed to update Knative Certificate", zap.Error(err))
+			c.Recorder.Eventf(existingCert, corev1.EventTypeWarning, "UpdateFailed",
+				"Failed to update Knative Certificate %s/%s: %v", desiredCert.Namespace, desiredCert.Name, err)
+			return err
+		}
+		c.Recorder.Eventf(existingCert, corev1.EventTypeNormal, "Updated",
+			"Updated Spec for Knative Certificate %s/%s", desiredCert.Namespace, desiredCert.Name)
+		return nil
+	}
+
+	return nil
+}
+
+func makeDesiredCertificate(namespace *corev1.Namespace, dnsName string) *v1alpha1.Certificate {
+	return &v1alpha1.Certificate{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      namespace.Name,
+			Namespace: namespace.Name,
+			Labels: map[string]string{
+				"networking.knative.dev/ownerNamespace": namespace.Name,
```

Comment:
nit: could we use `networking.OwnerNamespaceLabelKey` instead of "networking.knative.dev/ownerNamespace" ?